aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -935,6 +951,80 @@ std::tuple<Tensor&, Tensor&> symeig_out(Tensor& vals, Tensor& vecs, const Tensor
   return std::tuple<Tensor&, Tensor&>(vals, vecs);
 }
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ eig ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+template <typename scalar_t>
+static void apply_eig(Tensor& vals_, Tensor& vecs_, const Tensor& self, bool eigenvectors) {
+#ifndef USE_LAPACK
+  AT_ERROR("symeig: LAPACK library not found in compilation");
+#else
+  int n = self.size(0);
+  Tensor wi = at::empty({n}, self.dtype());
+  Tensor wr = at::empty({n}, self.dtype());
+  char jobvr = eigenvectors ? 'V' : 'N';
+  int lda = n;
+  int info;
+  if (n > 0) {
+    scalar_t wkopt;
+    scalar_t* vecs_data = eigenvectors ? vecs_.data<scalar_t>() : nullptr;
+    int ldvr = eigenvectors ? n : 1;
+
+    lapackEig<scalar_t>('N', jobvr, n, self.data<scalar_t>(), lda, wr.data<scalar_t>(), wi.data<scalar_t>(),
+      nullptr, 1, vecs_data, ldvr, &wkopt, -1, &info);
+
+    int lwork = (int)wkopt;
+    Tensor work = at::empty({lwork}, self.dtype());
+
+    lapackEig<scalar_t>('N', jobvr, n, self.data<scalar_t>(), lda, wr.data<scalar_t>(), wi.data<scalar_t>(),
+      nullptr, 1, vecs_data, ldvr, work.data<scalar_t>(), lwork, &info);
+  }
+
+  scalar_t* vals_data = vals_.data<scalar_t>();
+  scalar_t* wi_data = wi.data<scalar_t>();
+  scalar_t* wr_data = wr.data<scalar_t>();
+  for (int i = 0; i < n; i++) {
+    vals_data[2 * i] = wr_data[i];
+    vals_data[2 * i + 1] = wi_data[i];
+  }
+#endif
+}
+
+std::tuple<Tensor, Tensor> _eig_helper_cpu(const Tensor& self, bool eigenvectors) {
+  TORCH_CHECK(self.dim() == 2, "A should be 2 dimensional");
+  TORCH_CHECK(self.size(0) == self.size(1), "A should be square");
+  TORCH_CHECK(self.isfinite().all().item<bool>(), "A should not contain infs or NaNs");
+  int n = self.size(0);
+  Tensor vecs_;
+  if (eigenvectors) {
+    vecs_ = at::empty({n, n}, self.dtype()).transpose(0, 1);
+  } else {
+    vecs_ = at::empty({0}, self.dtype());
+  }
+  Tensor vals_ = at::empty({n, 2}, self.dtype());
+  Tensor self_ = cloneBatchedColumnMajor(self);
+
+  AT_DISPATCH_FLOATING_TYPES(self.scalar_type(), "eig_cpu", [&]{
+    apply_eig<scalar_t>(vals_, vecs_, self_, eigenvectors);
+  });
+  return std::tuple<Tensor&, Tensor&>(vals_, vecs_);
+}
+
+std::tuple<Tensor&, Tensor&> eig_out_cpu(Tensor& vals, Tensor& vecs, const Tensor& self, bool eigenvectors) {
```

Comment:
This function is accessible as user API, it means that you need to check that types of vals and vecs match the type of self (vecs should be checked only if they are needed). Also, please add test checking an error when wrong `out=` is supplied.